discovery/manager.go

```diff
@@ -0,0 +1,292 @@
+
+	"github.com/prometheus/prometheus/config"
+
+	"github.com/prometheus/prometheus/discovery/azure"
+	"github.com/prometheus/prometheus/discovery/consul"
+	"github.com/prometheus/prometheus/discovery/dns"
+	"github.com/prometheus/prometheus/discovery/ec2"
+	"github.com/prometheus/prometheus/discovery/file"
+	"github.com/prometheus/prometheus/discovery/gce"
+	"github.com/prometheus/prometheus/discovery/kubernetes"
+	"github.com/prometheus/prometheus/discovery/marathon"
+	"github.com/prometheus/prometheus/discovery/openstack"
+	"github.com/prometheus/prometheus/discovery/triton"
+	"github.com/prometheus/prometheus/discovery/zookeeper"
+)
+
+// Discoverer provides information about target groups. It maintains a set
+// of sources from which TargetGroups can originate. Whenever a discovery provider
+// detects a potential change, it sends the TargetGroup through its channel.
+//
+// Discoverer does not know if an actual change happened.
+// It does guarantee that it sends the new TargetGroup whenever a change happens.
+//
+// Discoverers should initially send a full set of all discoverable TargetGroups.
+type Discoverer interface {
+	// Run hands a channel to the discovery provider(consul,dns etc) through which it can send
+	// updated target groups.
+	// Must returns if the context gets canceled. It should not close the update
+	// channel on returning.
+	Run(ctx context.Context, up chan<- []*config.TargetGroup)
+}
+
+// type pool struct {
+// 	cancel func()
+// 	tgroups []*config.TargetGroup
+// }
+
+// NewManager is the Discovery Manager constructor
+func NewManager(ctx context.Context, logger log.Logger) *Manager {
+	return &Manager{
+		ctx:            ctx,
+		logger:         logger,
+		actionCh:       make(chan func()),
+		syncCh:         make(chan map[string][]*config.TargetGroup),
+		endpoints:      make(map[string]map[string][]*config.TargetGroup),
+		discoverCancel: []context.CancelFunc{},
+	}
+}
+
+// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+type Manager struct {
+	ctx            context.Context
+	logger         log.Logger
+	syncCh         chan map[string][]*config.TargetGroup // map[targetSetName]
+	actionCh       chan func()
+	discoverCancel []context.CancelFunc
+	endpoints      map[string]map[string][]*config.TargetGroup // map[targetSetName]map[providerName]
+}
+
+// Run starts the background processing
+func (m *Manager) Run() error {
+	for {
+		select {
+		case f := <-m.actionCh:
+			f()
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		}
+	}
+}
+
+// SyncCh returns a read only channel used by all DiscoveryProviders targetSet updates
+func (m *Manager) SyncCh() <-chan map[string][]*config.TargetGroup {
+	return m.syncCh
+}
+
+// ApplyConfig removes all running discovery providers and starts new ones using the provided config.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	err := make(chan error)
+	m.actionCh <- func() {
+		m.cancelDiscoverers()
```

Comment:
should we not be checking if `m.cancelDiscoverers` exists as we were before?

Reply:
not sure what you mean , how can not exist?

Reply:
Sorry, I guess should we be checking if the cancel func for `m.discoverCancel` exists, misread how this had changed. See my new comments below

Reply:
the `NewManager` constructor initialises to a zero value so no need for an extra check.

Reply:
facepalming, missed that `discoverCancel` was a field in the `Manager` struct

thanks! :)